homeassistant/components/combined_energy/coordinator.py

```diff
@@ -0,0 +1,136 @@
+"""DataUpdateCoordinator for the PVOutput integration."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from datetime import timedelta
+from typing import Generic, TypeVar
+
+from combined_energy import CombinedEnergy
+from combined_energy.exceptions import CombinedEnergyAuthError, CombinedEnergyError
+from combined_energy.helpers import ReadingsIterator
+from combined_energy.models import ConnectionStatus, DeviceReadings
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .backports import anext  # pylint: disable=redefined-builtin
+from .const import (
+    CONNECTIVITY_UPDATE_DELAY,
+    LOG_SESSION_REFRESH_DELAY,
+    LOGGER,
+    READINGS_INCREMENT,
+    READINGS_INITIAL_DELTA,
+    READINGS_UPDATE_DELAY,
+)
+
+_T = TypeVar("_T")
+
+
+class CombinedEnergyDataService(Generic[_T]):
```

Comment:
I'm a bit confused, why are you not just using a coordinator object? Feels like you're over complicating this. I could be missing something

Reply:
This is due to the underlaying API, which is actually multiple data services.

Since they all utilise the same basic pattern I provide that in a generic base class that each service inherits to provide the concrete implementation.